<launch>
    <arg name="use_sim_time" default="true"/>
    <arg name="log_level" default="warn"/>

    <arg name="resized_image" default="/sensing/camera/undistorted/image_raw"/>
    <arg name="resized_info" default="/sensing/camera/undistorted/camera_info"/>

    <!-- ========================================= -->
    <!-- ============= predict update ============ -->
    <!-- ========================================= -->
    <node pkg="modularized_particle_filter" exec="predictor" name="predictor" output="screen" args="--ros-args --log-level warn">
        <remap from="twist" to="/kalman/twist" />
        <remap from="initialpose" to="/initialpose" />
        <remap from="pose" to="/init/pose"/> 

        <param name="use_sim_time" value="$(var use_sim_time)"/>
        <param name="num_of_particles" value="500" />
        <param name="resampling_interval_seconds" value="1.5" />
        <param name="prediction_rate" value="50.0" />

        <param name="static_linear_covariance" value="0.05" />
        <param name="static_angular_covariance" value="0.005" />

        <param name="visualize" value="true"/>
        <remap from="/marker_array_particles" to="/weighted_particles/marker_array"/>
    </node>

    <!-- ========================================= -->
    <!-- ============ camera  correction ========= -->
    <!-- ========================================= -->
    <node name="camera_corrector" pkg="vector_map_visual_localizer" exec="camera_corrector_node" output="screen" args="--ros-args --log-level warn">
        <param name="use_sim_time" value="$(var use_sim_time)"/>
        <param name="image_size" value="800"/>
        <param name="max_range" value="40.0"/>
        <param name="gamma" value="5.0"/>
        <param name="score_offset" value="-128.0"/>
        <param name="max_raw_score" value="100000.0"/>
        <param name="min_prob" value="0.1"/>
        <param name="far_weight_gain" value="0.001"/>

        <remap from="/weighted_particles" to="/weighted_particles"/>
        <remap from="/lsd_cloud" to="/projected_lsd_cloud"/>
    </node>

    <!-- ========================================= -->
    <!-- ============= gnss correction =========== -->
    <!-- ========================================= -->
    <node name="gnss_corrector" pkg="vector_map_visual_localizer" exec="gnss_corrector_node" output="screen" args="--ros-args --log-level info">
        <param name="use_sim_time" value="$(var use_sim_time)"/>
        <param name="min_prob" value="0.1"/>
        <param name="flat_radius" value="2.0"/>
        <param name="sigma" value="4.0"/>
        <param name="rtk_enabled" value="$(var rtk_enabled)"/>

        <remap from="/weighted_particles" to="/weighted_particles"/>
    </node>
</launch>
